_posts/2021-09-29-signed-immutables-bug.md

```diff
@@ -0,0 +1,92 @@
+---
+layout: post
+published: true
+title: 'Signed Immutables Bug'
+date: '2021-09-29'
+author: Solidity Team
+category: Security Alerts
+---
+
+On September 28th, 2021, the Solidity team discovered that
+for immutable variables of a signed integer type shorter than 256 bits,
+sign extension (cleanup) of its value is not always properly performed.
+
+Inline assembly is required to access the value in its unclean state.
+The bug is present since the introduction of the
+`immutable` feature in Solidity 0.6.5 and is fixed in 0.8.9.
+
+We assigned the bug a severity of "very low".
+
+## Technical Details
+
+When immutable variables are assigned in Solidity during the construction phase, the value
+is stored in memory until the creation concludes. For types shorter than 32 bytes,
+these values were stored left-aligned within the 32 byte word and when the values were
+read again, they were shifted back accordingly, but no cleanup was performed on the values.
+In particular, the signextend operation was not performed on signed values.
+
+As an example, if you store ``-2`` in an ``int8``, it is stored in two's complement encoding
+as a single-byte value of ``0xfe``. Since the words in the EVM use 32 bytes, the value
+needs to be properly sign-extended before an operation like division is performed,
```

Comment:
```suggestion
needs to be properly sign-extended before an operation like signed-division is performed,
```

Reply:
Actually we don't do cleanup on signed division...

Reply:
We do. phew.